### PR TITLE
Wrappers for common usecases

### PR DESCRIPTION
**Pull request by anshprat**
This adds check_puppet, enable_puppet, set_config and check_config.
This also adds ability to add various keys at various levels (ex - enable_puppet at host level and enable_upgrade at role level) such that when getting all the available settings for a host,
the two are combined and returned.

### CODE REVIEW

**Comment by anshprat** [comment]
https://github.com/JioCloud/puppet-rjil/pull/768

File: jiocloud/orchestrate.py [hunk 1]
```diff
@@ -60,37 +60,52 @@ def manage_config(self, action_type, scope, data, name=None, action="set"):
         if not any(scope in s for s in ['global', 'role', 'host']):
             print "Invalid scope type: %s" % scope
             return False
-        if name is None:
-            if scope == 'global':
-                name_url = ''
-            else:
+        try:
+            cdata = data.split('=')
```
Comment (bodepd) [inline jiocloud/orchestrate.py]: why did you add an extra key here? It adds a lot of additional complexity, but I don't see what the use case is. I had originally intended for config_state to be synonymous with enable_puppet

Comment (anshprat) [inline jiocloud/orchestrate.py]: Basically this helps to add much more keys (and hence finer upgrade control) on each component. enable_puppet and enable_upgrade are two basic ones. But looking at the ceph upgrade plan, we need to have finer control like - is it ok to restart my mon now? can I restart my osd now?
So the idea to enable more keys is to enable more fine grain control over upgrade process. enable_upgrade will only say whether its ok to proceed with upgrade on not, but we also need similar flags for sub-parts of upgrade.

Comment (bodepd) [inline jiocloud/orchestrate.py]: ok, I don't mind that, but I'm maybe a little cautious of injecting keys for individual tasks related to upgrade things. I would much rather that we be able to solve problems in a more generic way (like using a session in consul to lock service restarts). I do understand that we might need to be able to use keys to coordinate things in the interim before we have more generic solutions.

Comment (anshprat) [inline jiocloud/orchestrate.py]: sessions definitely sound interesting. But we would need varying sessions across services (not the same session for a single node all services). The use case of ceph upgrade being - restart mon1 -> verify in quorum -> restart mon2..... -> after all mon are restarted and in quorum -> repeat same for all osd. Here, since the mon and osd are shared on 3 nodes, we need to revisit the stmon* twice for restarts - first for restart of mons and then after all mons are done, for osds.

**Comment by anshprat** [comment]
@bodepd btw, this is the consulr function lifted from another project and modified heavily for our purpose. essentially, I tried to rubify/puppetify the host_data + get_config to some extent.  https://github.com/anshprat/puppet-consulr/blob/manage_config/lib/puppet/parser/functions/consulr_kv.rb

File: jiocloud/orchestrate.py [hunk 2]
```diff
@@ -60,37 +60,52 @@ def manage_config(self, action_type, scope, data, name=None, action="set"):
...
+
+        return ret_dict
 
     def get_lookup_hash_from_hostname(self, name):
         m = re.search('(\w+)(\d+)(-.*)?', name)
         if m is None:
             print "Unexpected hostname format %s" % name
             return {}
-        return [['host', '/'+name], ['role', '/'+m.group(1)], ['global', '']]
+        return [['global', ''], ['role', '/'+m.group(1)], ['host', '/'+name] ]
```
Comment (bodepd) [inline jiocloud/orchestrate.py]: what was the motivation for changing the lookup order here? not sure if it matters, but this is potentially way less performant.

File: jiocloud/orchestrate.py [hunk 3]
```diff
@@ -60,37 +60,52 @@ def manage_config(self, action_type, scope, data, name=None, action="set"):
...
-                return result
-        return None
+#                 print result
+                for k in result.keys():
+                    ret_dict[k.rsplit('/',1)[-1]] = result[k]
+
+        return ret_dict
 
     def get_lookup_hash_from_hostname(self, name):
         m = re.search('(\w+)(\d+)(-.*)?', name)
```
Comment (bodepd) [inline jiocloud/orchestrate.py]: not related specifically to this patch, but this reqexp is wrong, in the case of more than one digit, it will only match the last one. We should change it to a character class

```
re.search('([a-z]+)?(\d+)(-.*)?', x)

```

File: jiocloud/orchestrate.py (hunk 2 above)
Comment (bodepd) [inline jiocloud/orchestrate.py]: ah, easier to test :)

Comment (anshprat) [inline jiocloud/orchestrate.py]: Actually the motivation was different. Lets say, we have enable_upgrade at role level and enable_puppet at host level. In the earlier model, it will return immediately after it gets the first valid entry. We want entries from all level and combine them as per order priority. So in the new model, we get keys from all levels in top down (increasing order of priority) and keep merging the resultant dicts/keys. So the net result/priority remains correct and we get keys from all levels.
